chunky/src/java/se/llbit/chunky/world/Chunk.java

```diff
@@ -174,29 +174,121 @@ private void loadSurface(@NotNull Tag data, ChunkData chunkData, int yMin, int y
+        Tag biomePaletteTag = getTagFromNames(section, "biomes\\palette");
+        if (biomePaletteTag.isList()) { //must be a post 21w39a section, so we load the 3d biomes
+          ListTag localBiomePalette = biomePaletteTag.asList();
+          int bitsPerBiome = 1;
+          if (localBiomePalette.size() > 2) {
+            bitsPerBiome = QuickMath.log2(QuickMath.nextPow2(localBiomePalette.size()));
+          }
+
+          Tag biomes = getTagFromNames(section, "biomes\\data");
+
+          if (biomes.isLongArray(bitsPerBiome)) {
+            int[] subpalette = new int[localBiomePalette.size()];
+            int paletteIndex = 0;
+            for (Tag item : localBiomePalette.asList()) {
+              Biome biome = Biomes.biomesByResourceLocation.get(((StringTag) item).getData());
+              if (biome == null) {
+//                Log.warnf("Missing biome %s! defaulting to first biome in the palette", ((StringTag) item).getData()); //TODO: re-enable once all 1.18 biomes are added
+                subpalette[paletteIndex] = 0; //technically unnecessary as it will default to 0, but here for clarity
+              } else {
+                subpalette[paletteIndex] = biomePalette.put(biome);
+              }
+              paletteIndex += 1;
+            }
+
+            BitBuffer buffer = new BitBuffer(biomes.longArray(), bitsPerBiome, true);
+            for (int biomeY = 0; biomeY < 4; biomeY++) {
+              int quartY = sectionMinBlockY + (biomeY << 2);
+              for (int biomeZ = 0; biomeZ < 4; biomeZ++) {
+                int quartZ = biomeZ << 2;
+                for (int biomeX = 0; biomeX < 4; biomeX++) {
+                  int quartX = biomeX << 2;
+
+                  int b0 = buffer.read();
+                  assert b0 < subpalette.length;
+                  for (int localY = 0; localY < 4; localY++) {
+                    for (int localZ = 0; localZ < 4; localZ++) {
+                      for (int localX = 0; localX < 4; localX++) {
+                        chunkData.setBiomeAt(quartX + localX, quartY + localY, quartZ + localZ, (byte) subpalette[b0]); //TODO: int biomes
+                      }
+                    }
+                  }
+                }
+              }
+            }
+          } else {
+            if (localBiomePalette.size() == 1) {
+              // biome palette exists, but data does not, we assume that section is entirely filled with one biome
+              Biome biome = Biomes.biomesByResourceLocation.get(((StringTag) localBiomePalette.get(0)).getData());
+              int biomeId;
+              if (biome == null) {
+//                Log.warnf("Missing biome %s! defaulting to first biome in the palette", ((StringTag) localBiomePalette.get(0)).getData());
+                biomeId = 0; //technically unnecessary as it will default to 0, but here for clarity
+              } else {
+                biomeId = biomePalette.put(biome);
+              }
```

Comment:
Same clarity, more self-explaining (imho):

```suggestion
              int biomeId = 0;
              if (biome != null) {
                biomeId = biomePalette.put(biome);
              }
```

Reply:
the reason I had this way is to allow for the log
We need a nice way to only take the first N warnings or something, as this spammed way too much